Explicit storage cost doc. (#48)
massalabs/massa-docs: technical-doc/storage-costs.rst

If you want to calculate the storage cost of your address in the ledger the formula is : address_size + balance_constant + bytecode_length + sum of (constants datastore key + value size) = 32 + 8 + bytecode.len() + sum_i(10 + datastore[i].len())) * 0.00025

> The storage costs are always paid by the address that call the ABI. If you are using ABI like set_bytecode you will be charged and also if you are using ABI like set_bycode_for you will be charged for the storage costs.

Reviewer: ```suggestion
The storage costs are always paid by the address that calls the ABI. For instance, if you are using the ABI `set_bytecode` or `set_bycode_for` you will be charged for the storage costs.
```
Either do something like this. Or explicitly list all the ABI that can end up charging you for storage costs.

You want to store your birth date that is 30 bytes long in your datastore so you need to send an operation that will create a key in your datastore using a SC. This operation will cost you in storage costs at least 0.00025 * (10 + 30) = 0.01 Massa.

Now you want to delete this entry on your datastore you will be refunded of the storage costs (0.02 Massa).

> In case you are interacting with a SC that will store the data on his datastore then he will be charged for the storage costs not you.

Reviewer: Don't get what you are trying to say here.

Author: I try to say that as a TX sender you will not pay the storage costs that SC will have (except if they use the coins you passed as parameter) I have reformulate a bit.

> In case you are interacting with a SC that will store the data on his datastore then the SC will pay for the storage costs not you as a TX sender (but he can use the coins you pass to the call to pay).

Reviewer: I think the english can be (greatly) improved here.

Author: I have made a new formulation.

Notes
-----

> In case of reimbursement of storage costs, the coins are not sent to the address that was charged but to the address that call the ABI to refund.

Reviewer: ```suggestion
In case of reimbursement of storage costs, the coins are not sent to the address that was charged but to the address that called the ABI to refund.
```
The sentence is still unclear. Especially this part: "that called the ABI to refund".

> In case of refund of the storage costs, after realeasing space, the address reimbursed is the one that calls the ABI, possibly different from the one that paid for this storage.

Author: As the developer wants but yes that's a possibility. It's up to him.

> If you are a SC developer and you want your users to pay for the storage costs of your smart contract you can use the coins that are passed by the `coins` parameter of CallSC.

Reviewer 2: This would imply that the contract needs to check how many coins the user sent, and if there are enough to execute the remaining instructions of the function then proceed to execution or revert the transaction, wouldn't it ? Otherwise the SC would just use its balance to execute the remaining instructions even the user didnt send enough coins, no ?

Author: It's possible that the SC check his balance as start of the code of his function and check it at the end with an assert and so if it fails then the execution is automatically revert by the core. We plan to add the number of coins used in the execution in read only call so that fronts of dApps can simulate and then fill the `coins` parameter the right way
